codetransformer/patterns.py

```diff
@@ -0,0 +1,289 @@
+
+    def __repr__(self):
+        return self._token.decode('utf-8')
+    __str__ = __repr__
+
+
+class modifier(meta):
+    """Marker class for modifier types.
+    """
+    pass
+
+
+@instance
+class star(modifier):
+    """Modifier that matches zero or more of a pattern.
+    """
+    _token = b'*'
+
+
+@instance
+class plus(modifier):
+    """Modifier that matches one or more of a pattern.
+    """
+    _token = b'+'
+
+
+@instance
+class option(modifier):
+    """Modifier that matches zero or one of a pattern.
+    """
+    _token = b'?'
+
+
+class matchnm(immutable, meta, defaults={'m': None}):
+    __slots__ = 'matchable', 'n', 'm'
+
+    def mcompile(self):
+        m = self.m
+        return (
+            self.matchable.mcompile() +
+            b'{' +
+            bytes(str(self.n), 'utf-8') +
+            (b'' if m is None else (b', ' + bytes(str(m), 'utf-8'))) +
+            b'}'
+        )
+
+    def __repr__(self):
+        return '{matchable}[{args}]'.format(
+            matchable=_prepr(self.matchable),
+            args=', '.join(map(str, filter(bool, (self.n, self.m)))),
+        )
+
+
+@instance
+class matchany(meta, matchable):
+    """Matchable that matches any instruction.
+    """
+    _token = b'.'
+
+    def __repr__(self):
+        return '...'
+
+
+class seq(immutable, matchable):
+    """A sequence of matchables to match in order.
+
+    Parameters
+    ----------
+    *matchables : iterable of matchable
+        The matchables to match against.
+    """
+    __slots__ = 'matchables',
+
+    def __new__(cls, *matchables):
+        if not matchables:
+            raise TypeError('cannot create an empty sequence')
+
+        if len(matchables) == 1:
+            return coerce_ellipsis(matchables[0])
+        return super().__new__(cls)
+
+    def __init__(self, *matchables):
+        self.matchables = tuple(map(coerce_ellipsis, matchables))
+
+    def mcompile(self):
+        return b''.join(map(mcompile, self.matchables))
+
+    def __repr__(self):
+        return '{cls}({args})'.format(
+            cls=type(self).__name__,
+            args=', '.join(map(_prepr, self.matchables))
+        )
+
+
+class or_(immutable, matchable):
+    """Logical or of multiple matchables.
+
+    Parameters
+    ----------
+    *matchables : iterable of matchable
+        The matchables to or together.
+    """
+    __slots__ = '*matchables',
+
+    def mcompile(self):
+        return b'(' + b'|'.join(map(mcompile, self.matchables)) + b')'
+
+    def __repr__(self):
+        return ' | '.join(map(_prepr, self.matchables))
+
+
+class not_(immutable):
+    """Logical not of a matchable.
+    """
+    __slots__ = 'matchable',
+
+    def mcompile(self):
+        matchable = self.matchable
+        if isinstance(matchable, (seq, or_, not_)):
+            return b'((?!(' + matchable.mcompile() + b')).)*'
+
+        return b'[^' + matchable.mcompile() + b']'
+
+    def __repr__(self):
+        return '~' + _prepr(self.matchable)
```

Comment:
The logic for prepr seems like it's tied to `not_`.  Should we just inline it here?